kube-proxy: internal config: add Linux and Windows section

aroradaman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Introduce Linux and Windows section for internal configuration of kube-proxy adhering to the v1alpha2 version specifications as detailed in https://kep.k8s.io/784. 
This also introduces WindowsRunAsService to the v1alpha1 configuration.
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-proxy Windows service control manager integration(--windows-service) is now configurable in v1alpha1 component configuration via `WindowsRunAsService` field
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aroradaman committed:
kube-proxy: internal config: add Windows section

Introduce Windows section for internal configuration of kube-proxy
adhering to the v1alpha2 version specifications as detailed in
https://kep.k8s.io/784. This also introduces WindowsRunAsService
to v1alpha1 configuration.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
/cc @danwinship @thockin

danwinship commented on pkg/proxy/apis/config/fuzzer/fuzzer.go:
```diff
@@ -51,6 +51,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 			if obj.Logging.Format == "" {
 				obj.Logging.Format = "text"
 			}
+			obj.Windows = kubeproxyconfig.KubeProxyWindowsConfiguration{}
```
You don't need to do that, and in fact you don't _want_ to do that; the `c.FuzzNoCustom(obj)` will already have fuzzed all of the scalar-typed fields (including `obj.Windows.RunAsService`), and this line will effectively undo that and force `RunAsService` to always be `false`.

Or did you add this to avoid unit test problems? It's possible that we're just going to have to add the new fields to the `v1alpha` config too...

aroradaman replied:
Yes, the unit test fails if I don't add that.
Should I simply add `WindowsRunAsService` to `v1alpha1.KubeProxyConfiguration`?

danwinship commented on pkg/proxy/apis/config/types.go:
```diff
@@ -22,6 +22,13 @@ import (
+// KubeProxyWindowsConfiguration contains windows-platform-related configuration details for the
+// Kubernetes proxy server that aren't specific to a particular backend
+type KubeProxyWindowsConfiguration struct {
+	// runAsService, if true, enables windows service control manager API integration.
```
```suggestion
	// runAsService, if true, enables Windows service control manager API integration.
```

likewise in the other comments

danwinship commented on pkg/proxy/apis/config/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,31 @@
+	if err := autoConvert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in, out, scope); err != nil {
+		return err
+	}
+	return nil
```
I assume you did it this way rather than just `return autoConvert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in, out, scope)` because other stuff is going to be added to this function in future commits?

aroradaman replied:
Yes, this was intentional.

danwinship commented on staging/src/k8s.io/kube-proxy/config/v1alpha1/types.go:
```diff
@@ -243,6 +243,9 @@ type KubeProxyConfiguration struct {
 	PortRange string `json:"portRange"`
+
+	// windowsRunAsService, if true, enables Windows service control manager API integration.
+	WindowsRunAsService bool `json:"windowsRunAsService"`
```
If you add `omitempty` then it won't show up in the output by default and you can revert the changes to the unit test data.

The alternative to this field would be to just add the whole `KubeProxyWindowsConfiguration` to v1alpha1. But I guess that would be weird since we won't need to add `KubeProxyLinuxConfiguration`. So this seems fine.

danwinship commented on pkg/proxy/apis/config/types.go:
```diff
@@ -165,6 +172,9 @@ type DetectLocalConfiguration struct {
 type KubeProxyConfiguration struct {
 	metav1.TypeMeta
 
+	// windows contains windows-related configuration options.
```
```suggestion
	// windows contains Windows-related configuration options.
```

aroradaman commented:
/test pull-kubernetes-e2e-capz-windows-master

aroradaman committed:
kube-proxy: internal config: add Linux section

Introduce Linux section for internal configuration of kube-proxy
adhering to the v1alpha2 version specifications as detailed in
https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
/test pull-kubernetes-node-e2e-containerd 

_Pod scheduling timeout_

danwinship commented on pkg/proxy/apis/config/v1alpha1/conversion.go:
```diff
@@ -28,6 +28,10 @@ func Convert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in
 	out.WindowsRunAsService = in.Windows.RunAsService
+	out.Conntrack = v1alpha1.KubeProxyConntrackConfiguration(in.Linux.Conntrack)
+	out.OOMScoreAdj = in.Linux.OOMScoreAdj
+	out.NFTables.MasqueradeAll = in.Linux.MasqueradeAll
+	out.IPTables.MasqueradeAll = in.Linux.MasqueradeAll
```
I'd probably make this conditional on `in.Mode` like with the other conversion func. (Ideally the mode-specific configs should all have default values when not using that mode...)

danwinship commented:
/lgtm
/approve

/assign @thockin
for API approval (adds a new field to the `v1alpha1` kube-proxy API corresponding to an existing command-line option / future `v1alpha2` config option)

This is part of #121830, which is reorganizing the internal version of `KubeProxyConfiguration` to make more sense, in preparation for the `v1alpha2` (and some day `v1`) API. This part just splits out the `Linux` and `Windows` sections, as detailed more in [the KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/784-kube-proxy-component-config#design-details). Needs API approval because of the new `v1alpha1` field (which, to be clear, we don't really intend for anyone to use; we just wanted it in `v1alpha2` and ran into round-tripping problems if we didn't add it to `v1alpha1` too).

thockin commented:
@danwinship you should add yourself in staging/src/k8s.io/kube-proxy/config/OWNERS

thockin commented:
I am not sure why we have all the component-config schemas in pkg/generated/openapi/zz_generated.openapi.go -- is it intentional or accidental?  Maybe @jpbetz or @liggitt knows?

/approve
